File: packages/catalog_core/lib/src/bundle.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:archive/archive_io.dart';

import 'entry.dart';
import 'fields.dart';
import 'store.dart';

/// Sync-by-messenger (ADR-0002 family): one file carries the sender's
/// full knowledge — entries plus the photos currently in use — through
/// WhatsApp, Signal, mail, anything that moves files. Importing runs
/// the same idempotent engine as every other transport, with the
/// bundle's own vector as causal context for conflict detection.
/// The bundle format this build writes and the highest it reads.
///
/// Format 2 (1.0.0, #74): entries carry the reminder flag and live in
/// `entries2.jsonl` beside a `format` marker file. Pre-1.0.0 readers
/// only look for `entries.jsonl`, so a v2 bundle imports as zero
/// entries there — a safe no-op instead of silently stripping the
/// flag and turning plans into facts. This reader still accepts the
/// old name for bundles written by pre-1.0.0 devices.
const bundleFormat = 2;

/// A bundle written by a newer app than this one reads.
class UnsupportedBundleFormat implements Exception {
  final int format;
  const UnsupportedBundleFormat(this.format);

  @override
  String toString() =>
      'Bundle format $format is newer than this app reads ($bundleFormat)';
}

class BundleResult {
  final int entriesIn;
  final int blobsIn;

  /// The entries actually new to this store — the import summary's input.
  final List<Entry> applied;

  const BundleResult(this.entriesIn, this.blobsIn,
      {this.applied = const []});

  @override
  String toString() => '$entriesIn entries + $blobsIn photos in';
}

/// Writes the bundle zip and returns its path. Private values and the
/// photos among them stay out unless [includePrivate] is set; what
/// identifies a cat or a clowder always travels.
String writeBundle(CatalogStore store, String path,
    {bool includePrivate = false}) {
  final entries =
      store.entriesSince(const {}, includePrivate: includePrivate);
  final jsonl = entries.map((e) => jsonEncode(e.toJson())).join('\n');
  final seen = <String>{};
  final hashes = <String>[];
  for (final entity in [...store.cats(), ...store.clowders()]) {
    for (final hash in store.images(entity.id)) {
      // A photo is a value like any other: withheld ones leave their
      // marker behind and their bytes at home.
      if (!includePrivate &&
          store.isFieldPrivate(entity.id, Keys.image(hash))) {
        continue;
      }
      if (seen.add(hash)) hashes.add(hash);
    }
  }
  writeZipStreaming(path, utf8.encode(jsonl),
      flagged: entries.any((e) => e.reminder),
      hashes: hashes,
      bytesOf: store.imageBytes);
  return path;
}

/// Writes a bundle zip one entry at a time: the entries deflated, each
/// photo copied as it is (a JPEG does not shrink), straight to the
/// file. The old way built the whole archive in memory and then a
/// second copy for the zip — two times every photo, on every automatic
/// backup; that is what killed the app on phones with many photos.
void writeZipStreaming(String path, List<int> jsonlBytes,
    {required bool flagged,
    required Iterable<String> hashes,
    required Uint8List? Function(String hash) bytesOf}) {
  // The encoder would create missing folders on its own; a bundle
  // belongs where the caller said and nowhere else, and callers rely on
  // "cannot be written" being an error (going back removes nothing).
  final parent = File(path).parent;
  if (!parent.existsSync()) {
    throw FileSystemException('No such directory', parent.path);
  }
  final encoder = ZipFileEncoder()..create(path);
  try {
    _addFormatAndEntries(encoder, jsonlBytes, flagged: flagged);
    for (final hash in hashes) {
      final bytes = bytesOf(hash);
      if (bytes == null) continue;
      encoder.addArchiveFile(
          ArchiveFile('blobs/$hash.jpg', bytes.length, bytes)
            ..compress = false);
    }
  } finally {
    encoder.closeSync();
  }
}

/// Writes a bundle of exactly [entries] plus the photos they mention.
///
/// Used by going back: the removed entries are written out before they
/// are deleted, so nothing is ever destroyed. The rows keep their
/// original (device, dseq) — this file comes home to the catalog it was
/// written from, where those numbers are what makes the restore exact
/// and a later deliberate re-import of the same material a no-op rather
/// than a doubled history.
String writeEntriesBundle(
    CatalogStore store, String path, List<Entry> entries) {
  final jsonl = entries.map((e) => jsonEncode(e.toJson())).join('\n');
  final seen = <String>{};
  final hashes = <String>[
    for (final e in entries)
      if (e.field.startsWith(Keys.imagePrefix) &&
          seen.add(e.field.substring(Keys.imagePrefix.length)))
        e.field.substring(Keys.imagePrefix.length)
  ];
  writeZipStreaming(path, utf8.encode(jsonl),
      flagged: entries.any((e) => e.reminder),
      hashes: hashes,
      bytesOf: store.imageBytes);
  return path;
}

/// A payload without a single flagged entry is byte-identical to the
/// pre-1.0.0 format, so it ships under the old name and stays
/// importable on 0.3.x. Only actual reminders force the v2 layout.
void _addFormatAndEntries(ZipFileEncoder encoder, List<int> jsonlBytes,
    {required bool flagged}) {
  if (!flagged) {
    encoder.addArchiveFile(
        ArchiveFile('entries.jsonl', jsonlBytes.length, jsonlBytes));
    return;
  }
  final format = utf8.encode('$bundleFormat');
  encoder.addArchiveFile(ArchiveFile('format', format.length, format));
  encoder.addArchiveFile(
      ArchiveFile('entries2.jsonl', jsonlBytes.length, jsonlBytes));
}

/// Imports a bundle file; unknown entries and missing photos land,
/// everything else is ignored.
BundleResult importBundle(CatalogStore store, String path) {
  // Read from the file as needed: a backup with hundreds of photos is
  // never in memory as a whole, let alone with every photo unpacked
  // beside it — that killed the app on restore.
  final input = InputFileStream(path);
  try {
    return _importArchive(store, ZipDecoder().decodeBuffer(input));
  } finally {
    input.close();
  }
}

/// Same import from in-memory bytes — QR share payloads never touch
/// disk (#40).
BundleResult importBundleBytes(CatalogStore store, List<int> zipBytes) =>
    _importArchive(store, ZipDecoder().decodeBytes(zipBytes));

BundleResult _importArchive(CatalogStore store, Archive archive) {
  final entries = <Entry>[];
  final blobFiles = <String, ArchiveFile>{};
  for (final file in archive.files) {
    if (!file.isFile) continue;
    if (file.name == 'format') {
      final declared = int.tryParse(
          utf8.decode(file.content as List<int>).trim());
      if (declared != null && declared > bundleFormat) {
        throw UnsupportedBundleFormat(declared);
      }
    } else if (file.name == 'entries2.jsonl' ||
        file.name == 'entries.jsonl') {
      for (final line in utf8.decode(file.content as List<int>).split('\n')) {
        if (line.trim().isEmpty) continue;
        entries.add(
            Entry.fromJson((jsonDecode(line) as Map).cast<String, dynamic>()));
      }
    } else if (file.name.startsWith('blobs/') && file.name.endsWith('.jpg')) {
      final hash = file.name.substring('blobs/'.length,
          file.name.length - '.jpg'.length);
      blobFiles[hash] = file;
    }
  }
  final writerVector = <String, int>{};
  for (final e in entries) {
    if (e.dseq > (writerVector[e.device] ?? 0)) {
      writerVector[e.device] = e.dseq;
    }
  }
  final applied = store.applyEntries(entries, senderVector: writerVector);
  final imported = applied.length;
  var blobsIn = 0;
  for (final MapEntry(key: hash, value: file) in blobFiles.entries) {
    // Anything the catalog knows of and does not hold — deleted photos
    // included, so an archive can be restored with its pictures. Each
    // photo is unpacked only now, stored, and let go before the next.
    if (!store.knowsImage(hash)) continue;
    if (store.imageBytes(hash) != null) continue;
    store.putBlob(hash, Uint8List.fromList(file.content as List<int>));
    file.clear();
    blobsIn++;
  }
  return BundleResult(imported, blobsIn, applied: applied);
}

File: desktop/crates/catlog-core/src/bundle.rs
//! Sync by messenger (ADR 0002 family): a `.catsync` zip carries the
//! sender's full knowledge, entries plus the photos in use. Format 2
//! keeps the entries in `entries2.jsonl` beside a `format` marker;
//! format 1 files, from before reminders, use `entries.jsonl`.

use std::collections::{BTreeMap, HashMap};
use std::io::{Read, Write};
use std::path::Path;

use zip::write::SimpleFileOptions;

use crate::Result;
use crate::catalog::Catalog;
use crate::entry::Entry;
use crate::error::Error;
use crate::folder::parse_lines;
use crate::signing::{ImportReport, KeyRecord, parse_keys};

/// The key list inside a bundle.
pub const KEYS_FILE: &str = "keys.json";

/// The bundle format this build writes and the highest it reads.
pub const BUNDLE_FORMAT: u32 = 2;

/// A photo larger than this is not one of ours and is skipped; an
/// entries file above its cap is refused before it is unpacked.
pub const MAX_BLOB_BYTES: u64 = 20 << 20;
pub const MAX_ENTRIES_BYTES: u64 = 64 << 20;

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct BundleResult {
    pub entries_in: usize,
    pub blobs_in: usize,
    pub applied: Vec<Entry>,
    /// Refused rows and the keys met.
    pub report: ImportReport,
    /// Fields with unresolved concurrent edits after this import.
    pub conflicts: Vec<(String, String)>,
}

impl BundleResult {
    /// The import as the fixture corpus records it.
    pub fn to_json(&self) -> serde_json::Value {
        serde_json::json!({
            "entriesIn": self.entries_in,
            "blobsIn": self.blobs_in,
            "conflicts": self.conflicts.iter().map(|(e, f)| vec![e.clone(), f.clone()]).collect::<Vec<_>>(),
        })
    }
}

impl Catalog {
    /// Writes the bundle zip: this Catalog's full knowledge plus the
    /// photos in use. Private values and the photos among them stay out
    /// unless `include_private`.
    pub fn write_bundle(&self, path: &Path, include_private: bool) -> Result<()> {
        let entries = self.entries_since(&BTreeMap::new(), include_private)?;
        let hashes = self.live_images(include_private)?;
        self.write_zip(path, &entries, &hashes)
    }

    /// Writes a bundle of exactly `entries` plus the photos they mention.
    /// Used by going back: the rows keep their original (device, dseq),
    /// so the file comes home exact.
    pub fn write_entries_bundle(&self, path: &Path, entries: &[Entry]) -> Result<()> {
        let mut seen = std::collections::HashSet::new();
        let hashes: Vec<String> = entries
            .iter()
            .filter_map(|e| e.field.strip_prefix(crate::keys::IMAGE_PREFIX))
            .filter(|h| seen.insert(h.to_string()))
            .map(String::from)
            .collect();
        self.write_zip(path, entries, &hashes)
    }

    fn write_zip(&self, path: &Path, entries: &[Entry], hashes: &[String]) -> Result<()> {
        let mut jsonl = String::new();
        for (i, e) in entries.iter().enumerate() {
            if i > 0 {
                jsonl.push('\n');
            }
            jsonl.push_str(&serde_json::to_string(&e.wire())?);
        }
        let flagged = entries.iter().any(|e| e.reminder);
        let file = std::fs::File::create(path).map_err(|e| Error::io(path, e))?;
        let mut zip = zip::ZipWriter::new(file);
        let deflated = SimpleFileOptions::default()
            .compression_method(zip::CompressionMethod::Deflated)
            .last_modified_time(zip::DateTime::default());
        let stored = SimpleFileOptions::default()
            .compression_method(zip::CompressionMethod::Stored)
            .last_modified_time(zip::DateTime::default());
        // A payload without a flagged entry is byte-identical to the
        // pre-1.0.0 format and ships under the old name.
        if flagged {
            zip.start_file("format", deflated)?;
            zip.write_all(BUNDLE_FORMAT.to_string().as_bytes())
                .map_err(|e| Error::io(path, e))?;
            zip.start_file("entries2.jsonl", deflated)?;
        } else {
            zip.start_file("entries.jsonl", deflated)?;
        }
        zip.write_all(jsonl.as_bytes())
            .map_err(|e| Error::io(path, e))?;
        // The keys the entries were signed with; a reader from before
        // ignores the file, as it ignores the signatures.
        zip.start_file(KEYS_FILE, deflated)?;
        zip.write_all(serde_json::to_string(&self.key_records()?)?.as_bytes())
            .map_err(|e| Error::io(path, e))?;
        for hash in hashes {
            let Some(bytes) = self.image_bytes(hash) else {
                continue;
            };
            zip.start_file(format!("blobs/{hash}.jpg"), stored)?;
            zip.write_all(&bytes).map_err(|e| Error::io(path, e))?;
        }
        zip.finish()?;
        Ok(())
    }

    /// Imports a bundle file; unknown entries and missing photos land,
    /// everything else is ignored.
    pub fn import_bundle(&mut self, path: &Path) -> Result<BundleResult> {
        let file = std::fs::File::open(path).map_err(|e| Error::io(path, e))?;
        let mut archive = zip::ZipArchive::new(file)?;
        let mut entries: Vec<Entry> = Vec::new();
        let mut keys: Vec<KeyRecord> = Vec::new();
        let mut blobs: HashMap<String, usize> = HashMap::new();
        for i in 0..archive.len() {
            let mut f = archive.by_index(i)?;
            if f.is_dir() {
                continue;
            }
            let name = f.name().to_string();
            if name.starts_with("blobs/") {
                if f.size() > MAX_BLOB_BYTES {
                    continue;
                }
            } else if f.size() > MAX_ENTRIES_BYTES {
                return Err(Error::BundleTooLarge);
            }
            if name == "format" {
                let mut text = String::new();
                f.read_to_string(&mut text)
                    .map_err(|e| Error::io(path, e))?;
                if let Ok(declared) = text.trim().parse::<u32>()
                    && declared > BUNDLE_FORMAT
                {
                    return Err(Error::UnsupportedBundleFormat(declared));
                }
            } else if name == KEYS_FILE {
                let mut text = String::new();
                f.read_to_string(&mut text)
                    .map_err(|e| Error::io(path, e))?;
                keys = parse_keys(&text);
            } else if name == "entries2.jsonl" || name == "entries.jsonl" {
                let mut text = String::new();
                f.read_to_string(&mut text)
                    .map_err(|e| Error::io(path, e))?;
                entries.extend(parse_lines(&text).unwrap_or_default());
            } else if let Some(rest) = name.strip_prefix("blobs/")
                && let Some(hash) = rest.strip_suffix(".jpg")
            {
                blobs.insert(hash.to_string(), i);
            }
        }
        let mut report = ImportReport::default();
        let writer_vector = crate::folder::writer_vector(&entries);
        let applied =
            self.apply_entries_from(entries, &writer_vector, Some(&keys), None, &mut report)?;
        let mut blobs_in = 0;
        for (hash, index) in blobs {
            if !self.knows_image(&hash)? || self.image_bytes(&hash).is_some() {
                continue;
            }
            let mut f = archive.by_index(index)?;
            let mut bytes = Vec::with_capacity(f.size() as usize);
            f.read_to_end(&mut bytes).map_err(|e| Error::io(path, e))?;
            if self.put_blob(&hash, &bytes).is_ok() && self.image_bytes(&hash).is_some() {
                blobs_in += 1;
            }
        }
        let conflicts = self.conflicts()?;
        Ok(BundleResult {
            entries_in: applied.len(),
            blobs_in,
            applied,
            report,
            conflicts,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use zip::write::SimpleFileOptions;

    fn bundle(dir: &Path, files: &[(&str, &[u8])]) -> std::path::PathBuf {
        let path = dir.join("b.catsync");
        let mut zip = zip::ZipWriter::new(std::fs::File::create(&path).unwrap());
        for (name, bytes) in files {
            zip.start_file(*name, SimpleFileOptions::default()).unwrap();
            zip.write_all(bytes).unwrap();
        }
        zip.finish().unwrap();
        path
    }

    #[test]
    fn a_written_bundle_comes_back_whole_and_a_plan_forces_format_two() {
        let dir = tempfile::tempdir().unwrap();
        let mut a = Catalog::open_with_device(&dir.path().join("a"), "aaaa").unwrap();
        a.set_author("Ada").unwrap();
        let seeded = a.all_entries().unwrap().len();
        a.create_cat("cat:m", "Miezi", None, "cat").unwrap();
        let photo = a.add_image("cat:m", b"jpeg bytes").unwrap();
        let path = dir.path().join("a.catsync");
        a.write_bundle(&path, false).unwrap();
        let names: Vec<String> = {
            let z = zip::ZipArchive::new(std::fs::File::open(&path).unwrap()).unwrap();
            z.file_names().map(String::from).collect()
        };
        assert!(names.contains(&"entries.jsonl".to_string()));
        assert!(names.contains(&KEYS_FILE.to_string()));
        assert!(!names.contains(&"format".to_string()));

        let mut b = Catalog::open_with_device(&dir.path().join("b"), "bbbb").unwrap();
        let r = b.import_bundle(&path).unwrap();
        assert_eq!((r.entries_in, r.blobs_in), (seeded + 4, 1));
        assert_eq!(r.report.new_keys[0].record.device, "aaaa");
        assert!(b.import_bundle(&path).unwrap().report.is_empty());
        assert_eq!(b.image_bytes(&photo).as_deref(), Some(&b"jpeg bytes"[..]));
        assert_eq!(b.import_bundle(&path).unwrap().entries_in, 0);

        a.append_at(
            "cat:m",
            "f:vet",
            Some("shots"),
            Some("2026-03-01T00:00:00Z"),
            true,
        )
        .unwrap();
        a.write_bundle(&path, false).unwrap();
        let names: Vec<String> = {
            let z = zip::ZipArchive::new(std::fs::File::open(&path).unwrap()).unwrap();
            z.file_names().map(String::from).collect()
        };
        assert!(names.contains(&"entries2.jsonl".to_string()));
        assert!(names.contains(&"format".to_string()));
        assert_eq!(b.import_bundle(&path).unwrap().entries_in, 1);
        assert!(
            a.write_bundle(&dir.path().join("nowhere").join("x.catsync"), false)
                .is_err()
        );
    }

    #[test]
    fn a_newer_format_is_refused_and_an_unknown_photo_ignored() {
        let dir = tempfile::tempdir().unwrap();
        let mut c = Catalog::open(&dir.path().join("cat")).unwrap();
        let path = bundle(dir.path(), &[("format", b"3"), ("entries2.jsonl", b"")]);
        assert!(matches!(
            c.import_bundle(&path),
            Err(Error::UnsupportedBundleFormat(3))
        ));
        let path = bundle(
            dir.path(),
            &[
                ("format", b"2"),
                ("entries2.jsonl", br#"{"device":"w","dseq":1,"entity":"cat:a","field":"name","value":"M","date":"2026-01-01T00:00:00Z","author":"A","recorded":"2026-01-01T00:00:00Z","reminder":true}"#),
                ("blobs/ffff.jpg", b"stray"),
                ("blobs/", b""),
            ],
        );
        let r = c.import_bundle(&path).unwrap();
        assert_eq!(r.entries_in, 1);
        assert_eq!(r.blobs_in, 0);
        assert!(
            c.import_bundle(&dir.path().join("missing.catsync"))
                .is_err()
        );
    }
}
